[CSS-anchor-position-1] Fine-grained invalidation

anttijk opened the pull request:
#### 5daa772e687cbea74995e542e6412ffb7bddb1f3
<pre>
[CSS-anchor-position-1] Fine-grained invalidation
<a href="https://bugs.webkit.org/show_bug.cgi?id=287552">https://bugs.webkit.org/show_bug.cgi?id=287552</a>
<a href="https://rdar.apple.com/144681624">rdar://144681624</a>

Reviewed by Alan Baradlay.

Track when anchors move and invalidate anchored element style only when needed.
Before this we just invalidated all anchored elements unconditionally.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invalidateForAnchorRectChange):

Add a style invalidation function for this.

* Source/WebCore/dom/Element.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::layout):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):

Take care to also clear the offsets if they are not being applied.

(WebCore::Style::AnchorPositionEvaluator::makeAnchorPositionedForAnchorMap):

Make a reverse map.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForLayoutDependencies):
(WebCore::Style::Scope::invalidateForContainerDependencies):
(WebCore::Style::Scope::invalidateForAnchorDependencies):

Track the positions of anchors, invalidate if they have moved.

(WebCore::Style::Scope::resetAnchorPositioningStateBeforeStyleResolution):

Reset the transient state only.

(WebCore::Style::Scope::updateAchorPositioningStateAfterStyleResolution):

Update the snapshotted offsets.
Get rid of items that are not anchored anymore.

(WebCore::Style::Scope::updateQueryContainerState): Deleted.
(WebCore::Style::Scope::clearAnchorPositioningState): Deleted.
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::TreeResolver):
(WebCore::Style::TreeResolver::~TreeResolver):

Canonical link: <a href="https://commits.webkit.org/290291@main">https://commits.webkit.org/290291@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f403110295534fa8a3c72cddd94c7adf89d85c4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89554 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9081 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44413 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94546 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40321 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91606 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9470 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17360 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68985 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26642 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92555 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7272 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81278 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49350 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35662 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39427 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36665 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96374 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16736 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77853 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16991 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77081 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77168 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19046 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21608 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9903 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16749 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22063 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16490 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19941 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18272 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/StyleScope.cpp:
```diff
@@ -1033,16 +1082,23 @@ Element* hostForScopeOrdinal(const Element& element, ScopeOrdinal scopeOrdinal)
+}
 
-    m_anchorPositionedStates.clear();
+void Scope::updateAchorPositioningStateAfterStyleResolution()
```
```suggestion
void Scope::updateAnchorPositioningStateAfterStyleResolution()
```

anttijk committed:
[CSS-anchor-position-1] Fine-grained invalidation
https://bugs.webkit.org/show_bug.cgi?id=287552
rdar://144681624

Reviewed by Alan Baradlay.

Track when anchors move and invalidate anchored element style only when needed.
Before this we just invalidated all anchored elements unconditionally.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invalidateForAnchorRectChange):

Add a style invalidation function for this.

* Source/WebCore/dom/Element.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::layout):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):

Take care to also clear the offsets if they are not being applied.

(WebCore::Style::AnchorPositionEvaluator::makeAnchorPositionedForAnchorMap):

Make a reverse map.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForLayoutDependencies):
(WebCore::Style::Scope::invalidateForContainerDependencies):
(WebCore::Style::Scope::invalidateForAnchorDependencies):

Track the positions of anchors, invalidate if they have moved.

(WebCore::Style::Scope::resetAnchorPositioningStateBeforeStyleResolution):

Reset the transient state only.

(WebCore::Style::Scope::updateAchorPositioningStateAfterStyleResolution):

Update the snapshotted offsets.
Get rid of items that are not anchored anymore.

(WebCore::Style::Scope::updateQueryContainerState): Deleted.
(WebCore::Style::Scope::clearAnchorPositioningState): Deleted.
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::TreeResolver):
(WebCore::Style::TreeResolver::~TreeResolver):

Canonical link: https://commits.webkit.org/290291@main
